### PR TITLE
fix:implement Display trait for ScalarRefImp&&ScalarImpl manually 

### PR DESCRIPTION
**Pull request by ZENOTME**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

We use to_string() to translate a value in the database(ScalarRefImpl) to a TEXT format result. 

We used to implement Display trait in macro (using parse_display). But this will decrease the flexibility. Such as for the `bool` type, we can't directly use `bool.to_string` as TEXT format. We have to covert it to 't' or 'f'.
`(DataType::Boolean, ScalarRefImpl::Bool(b)) => if b { "t" } else { "f" }.into()`

This PR remove Display in macro, implement Display trait manually. 

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/common/src/types/mod.rs [hunk 1]
```diff
@@ -734,6 +734,54 @@ pub fn display_datum_ref(d: DatumRef<'_>) -> String {
...
+impl std::fmt::Display for ScalarImpl {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ScalarImpl::Int16(v) => write!(f, "{}", v),
+            ScalarImpl::Int32(v) => write!(f, "{}", v),
+            ScalarImpl::Int64(v) => write!(f, "{}", v),
+            ScalarImpl::Float32(v) => write!(f, "{}", v),
+            ScalarImpl::Float64(v) => write!(f, "{}", v),
+            ScalarImpl::Utf8(v) => write!(f, "{}", v),
+            ScalarImpl::Bool(v) => write!(f, "{}", v),
```
Comment (wangrunji0408) [inline src/common/src/types/mod.rs:746]: There should be `t` / `f` as well?

**Comment by jon-chuang** [comment]
> We use to_string() to translate a value in the database(ScalarRefImpl) to a TEXT format result.

Why is this a problem?

Do we lose information about type? Meaning, now we cannot distinguish between a ScalarImpl::Int16 and ..::Int64. IMO that's not a good change.

**Comment by xiangjinwu** [comment]
There have been some misuses of `Display` on `Scalar`, as shown by examples in https://github.com/risingwavelabs/risingwave/issues/4769#issuecomment-1290908459. Actually PostgreSQL is not consistent itself and uses different formats in different contexts:
* boolean
  * `true`: `true::varchar`, `array[true]::varchar[]`
  * `t`: `true`, `row(true)`, `row(true)::varchar`, `array[true]`, `array[true]::varchar`
* null:
  * (empty): `null`, `row(null)`
  * `NULL`: `array[null]`
  * We defined `type Datum = Option<Scalar>` so it was not possible to use `Display`.
* timestamp with time zone
  * It is currently also `Scalar::Int64`. Even if we introduce `Scalar::Timestampz`, it is still incorrect to simply `Display` the `Scalar`. The proper formatting has to take an extra parameter of session TimeZone.

To summarize my thoughts, any user-facing "to-string" that needs to be PostgreSQL-compatible should implement their own logic and avoid `Display`. And I currently do not have a good idea of what `Display` should be. Maybe it should be removed and replaced by several dedicated utils eventually.

**Comment by neverchanje** [comment]
'true': `true::varchar, array[true]::varchar[]`
It sounds literally crazy... What if I use "t" instead of "true"? Will it break anything? I just tested the behavior of cockroach:

boolean
- true: true, true::varchar, array[true]::varchar[], array[true]
- t:, row(true), row(true)::varchar, array[true]::varchar

There're a couple of differences but it seems cockroach never encountered any real problems because of the inconsistencies.

**Comment by ZENOTME** [comment]
> To summarize my thoughts, any user-facing "to-string" that needs to be PostgreSQL-compatible should implement their own logic and avoid Display. And I currently do not have a good idea of what Display should be. Maybe it should be removed and replaced by several dedicated utils eventually.

- I add the ToText trait responsible to convert a scalar type to text format
- Now we can implement own logic in ToText trait, for example:
```
impl ToText for bool {
    fn to_text(&self) -> String {
        if *self {
            "t".to_string()
        } else {
            "f".to_string()
        }
    }
}
```
- To simplify, for type whose text format is as same as the to_string(), I directly call to_string() to implement it.(Which means that we still use Display for these type). But we can implement their own ToText logic different than to_string() if we needed.
```
implement_using_to_string! {
    { i16 },
    { i32 },
    { i64 },
    { String },
    { &str },
    { crate::types::OrderedF32 },
    { crate::types::OrderedF64 },
    { crate::types::Decimal },
    { crate::types::IntervalUnit },
    { crate::types::NaiveDateWrapper },
    { crate::types::NaiveDateTimeWrapper },
    { crate::types::NaiveTimeWrapper }
}
```

File: src/frontend/src/handler/util.rs [hunk 2]
```diff
@@ -87,9 +87,9 @@ fn pg_value_format(data_type: &DataType, d: ScalarRefImpl<'_>, format: bool) ->
     // format == true means BINARY format
     if !format {
         match (data_type, d) {
-            (DataType::Boolean, ScalarRefImpl::Bool(b)) => if b { "t" } else { "f" }.into(),
+            // TODO: introduce ScalarRefImpl::TimestampZ
```
Comment (xiangjinwu) [inline src/frontend/src/handler/util.rs:90]: Have not decided whether to introduce `ScalarRefImpl::Timestampz` yet. It actually prevents some mis-assumption that `Scalar` contains enough type information, where `DataType` actually needs to be passed as argument together.

File: src/common/src/array/list_array.rs [hunk 3]
```diff
@@ -503,6 +504,42 @@ impl Debug for ListRef<'_> {
     }
 }
 
+impl ToText for ListRef<'_> {
+    fn to_text(&self) -> String {
+        iter_elems_ref!(self, it, {
+            format!(
+                "{{{}}}",
+                it.format_with(",", |datum_ref, f| {
+                    let s = datum_ref.to_text();
```
Comment (xiangjinwu) [inline src/common/src/array/list_array.rs:513]: Feels like duplicated with `Display` below except the usage of `to_text` here. However, there is no need to refactor and share code between these two. I actually found it is not hard to remove `Display` (as suggested in my previous comment). So this intermediate state is acceptable from my point of view.

**Review by xiangjinwu — APPROVED** [review]

File: src/common/src/types/decimal.rs [hunk 4]
```diff
@@ -31,6 +33,17 @@ pub enum Decimal {
     NegativeInf,
 }
 
+impl ToText for crate::types::Decimal {
+    fn to_text(&self) -> String {
+        match self {
```
Comment (xiangjinwu) [inline src/common/src/types/decimal.rs:38]: Why not just call `self.to_string()`? (And same for interval)

File: src/common/src/array/list_array.rs [hunk 5]
```diff
@@ -503,16 +505,13 @@ impl Debug for ListRef<'_> {
     }
 }
 
-impl Display for ListRef<'_> {
-    // This function will be invoked when pgwire prints a list value in string.
-    // Refer to PostgreSQL `array_out` or `appendPGArray`.
```
Comment (xiangjinwu) [inline src/common/src/array/list_array.rs:508]: Keep this comment

File: src/common/src/types/decimal.rs (hunk 4 above)
Comment (ZENOTME) [reply src/common/src/types/decimal.rs:38]: This can separate them if we want to remove to_string in the future.

Comment (xiangjinwu) [reply src/common/src/types/decimal.rs:38]: By that time we can just move code from `Display` into `ToText`. The interface of `ToText` remains the same and no updates to callers are needed. Keeping two duplicated snippets would be hard to maintain.

**Review by xiangjinwu — APPROVED** [review]
